### ISSUE TITLE
Specifying types in bulk requests is deprecated

### ISSUE THREAD

**Issue by fdartayre** [logstash-plugins/logstash-output-elasticsearch#915]
**Version:** logstash-output-elasticsearch 10.2.3 (shipping with Logstash 7.5.2)

[Starting from Elasticsearch 7.0](https://www.elastic.co/guide/en/elasticsearch/reference/current/removal-of-types.html#_document_apis), types should also no longer appear in the body of bulk requests.

Including a type causes a deprecation warning to get logged in Elasticsearch:
```
[2020-01-31T10:58:52,319][WARN ][o.e.d.a.b.BulkRequestParser] [hostname] [types removal] Specifying types in bulk requests is deprecated.
```
The plugin shouldn't use types anymore.

**Comment by andsel** [comment]
With commit https://github.com/logstash-plugins/logstash-output-elasticsearch/commit/c93abd1e6255d5ba519e942a85ac708c05d472a7 present in version `10.2.2` type was removed but with https://github.com/logstash-plugins/logstash-output-elasticsearch/commit/39a4db99008704553afb885a8ad9f0fbd44984fb was added because the `_monitoring/bulk` operation required it to continue work. With the LS 7.7 the end point should be switched and then removed

**Comment by rolffokkens** [comment]
Currently at LS 7.10, end the ES (also 7.10) logs are flouding my filesystem. No idea what the previous comment means.

**Comment by barosch78** [comment]
> Currently at LS 7.10, end the ES (also 7.10) logs are flouding my filesystem. No idea what the previous comment means. The flouding ("[types removal] Specifying types in bulk requests is deprecated.") started after upgrading ES from 7.9 to 7.10.

I have the same problem after upgrading to 7.10

**Comment by nosferatu314** [comment]
Same problem here. Flooding can be stopped by changing log level of deprecation warnings.

From: https://www.elastic.co/guide/en/elasticsearch/reference/current/logging.html

You can disable it in the config/log4j2.properties file by setting the deprecation log level to error like this:
```
logger.deprecation.name = org.elasticsearch.deprecation
logger.deprecation.level = error
```

This is merely a workaround and I would appreciate a fix.

**Comment by MakoWish** [comment]
Any updates on this issue?

**Comment by andsel** [comment]
Could you share some information about this problem please, which `logstash-output-elasticseach` configuration are you using?Did you remove the `type` from your config definition when pushing data to Elasticsearch?

**Comment by MakoWish** [comment]
Hi @andsel,

We are sending Beats data through Logstash for some additional data massaging. The entire stack is 7.10.2. That includes Beats, Logstash, Elasticsearch, Kibana...

Beats output:
```
output.logstash:
  hosts: [
    "LS1.Contoso.com:5044",
    "LS2.Contoso.com:5044",
    "LS3.Contoso.com:5044",
    "LS4.Contoso.com:5044"
  ]
  ssl.certificate: "${path.config}\\certs\\client.crt"
  ssl.key: "${path.config}\\certs\\client.key"
  ssl.key_passphrase: "${SECRET_PASS}"
```

Nearly all our data is from Beats modules, using the included Beats index templates. None of our non-Beats sources have a "_type" field explicitly added. It seems either Beats or Logstash still seem to add a field:value of "_type" : "_doc" to every event ingested. Basically, I am not adding a "_type" field anywhere, so it seems to be part of the stack itself. 

Would throwing something in the Logstash pipeline like this do the trick? 
```
mutate {
    remove_field => ["_type"]
}
```

Just to be 100% sure I wasn't adding a "_type" field anywhere, I ran `sudo cat /etc/logstash/conf.d/*.conf | grep -i "_type"`. The only things found were fields from grok match patterns (ie. `EVENT_TYPE = %{WORD:[event][type]}` matched).

**Comment by MakoWish** [comment]
UPDATE: I tried removing the "_type" field in my pipeline as noted above, but it is still in every event.

**Comment by andsel** [comment]
@MakoWish are you indexing Elasticsearch's deprecation logs in Elasticsearch itself?

**Comment by andsel** [comment]
It's assigned by `logstash-output-elasticsearh` plugins at:
https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/3769e2d88612b74023d717b663bd0ee3c4287166/lib/logstash/outputs/elasticsearch.rb#L328

with value:
https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/3769e2d88612b74023d717b663bd0ee3c4287166/lib/logstash/outputs/elasticsearch.rb#L388

**Comment by MakoWish** [comment]
@andsel,

Thank you for the response! Will that be disabled automatically with v8? Or is there a way to disable it now? I would like to stop the literally millions of messages.

**Comment by andsel** [comment]
A reproducer for this:
- download Elasticsearch `7.10.0` tar.gz, unpack and run Elasticsearch `bin/elasticsearch`
- download Logstasn `7.10.0`, unpack it
- create a pipeline config (name it `test_pipeline.conf`) with content:
```
input {
	generator {
		message => "{\"name\": \"John Doe\"}"
		count => 20000
	}
}

output {
	elasticsearch {
		index => "test_bulk_index"
	}
}
```
- run `bin/logstash -f test_pipeline.conf`
- check the `logs/elasticserach_deprecation.log` increasing

**Comment by andsel** [comment]
Logstash output Elasticsearch plugin uses the `type` when connect to a cluster of version `< 8.0`. Working on PR #994 to avoid it's usage also when connecting to Elasticsearch 7 and user doesn't specify `type` setting in plugin configuration.

As a fix on Elasticsearch side, instead of changing the level of deprecation logger:
- search for `Deprecation -  old style pattern` section in `config/log4j2.properties` in Elasticsearch
- append this config `appender.deprecation_rolling_old.filter.rate_limit.type = RateLimitingFilter`

**Comment by andsel** [comment]
Reference to Elasticsearch issue https://github.com/elastic/elasticsearch/issues/69188

**Comment by andsel** [comment]
This has been fixed with version `10.8.3` of `logstash-output-elasticsearch`
